File: Daily/sbgo/Leetcode697-数组的度.go
package sbgo

func findShortestSubArray(nums []int) int {
	mp := make(map[int][]int)
	for i, num := range nums {
		if array, has := mp[num]; has {
			array[0]++
			array[2] = i
		} else {
			mp[num] = []int{1, i, i}
		}
	}
	ans, maxnum := 0, 0
	for _, array := range mp {
		if array[0] > maxnum {
			ans = array[2] - array[1] + 1
			maxnum = array[0]
		} else if maxnum == array[0] {
			if ans > array[2]-array[1]+1 {
				ans = array[2] - array[1] + 1
			}
		}
	}
	return ans
}

File: Leetcode/PY/Leetcode697-数组的度.py
from typing import List
from collections import defaultdict
class Solution:
    def findShortestSubArray(self, nums: List[int]) -> int:
        mp = defaultdict(list)
        for i,num in enumerate(nums):
            if num in mp:
                mp[num][0] += 1
                mp[num][2] = i
            else:
                mp[num] = [1,i,i]
        ans = 0 
        maxnums = 0
        for time,start,end in mp.values():
            if maxnums < time:
                maxnums = time
                ans = end - start + 1
            elif maxnums == time:
                #海象运算符号，避免两次运算
                if ans > (tmp := end - start + 1):
                    ans = tmp
        return ans